files/pull-image.sh

```diff
@@ -0,0 +1,23 @@
+#!/usr/bin/env bash
+
+img=$1
+region=$(echo "${img}" | cut -f4 -d ".")
+MAX_RETRIES=3
+
+function retry() {
+    local rc=0
+    for attempt in `seq 0 $MAX_RETRIES`; do
+        rc=0
+        [[ $attempt -gt 0 ]] && echo "Attempt $attempt of $MAX_RETRIES" 1>&2
+        $@
+        rc=$?;
+        [[ $rc -eq 0 ]] && break
+        [[ $attempt -eq $MAX_RETRIES ]] && return exit $rc
+        local jitter=$((1 + RANDOM % 10))
+        local sleep_sec="$(( $(( 5 << $((1+$attempt)) )) + $jitter))"
+        sleep $sleep_sec
+    done
+}
+
+ecr_password=$(retry aws ecr get-login-password --region $region)
+retry sudo ctr --namespace k8s.io image pull "${img}" --user AWS:${ecr_password}
```

Comment:
With no Instance Profile configured it will get error like this ( with debug mode ON )

```
2022-10-16T01:08:59+08:00:     amazon-ebs: Unable to locate credentials. You can configure credentials by running "aws configure".
2022-10-16T01:08:59+08:00:     amazon-ebs: ++ retry aws ecr get-login-password --region us-west-2
2022-10-16T01:08:59+08:00:     amazon-ebs: ++ local rc=0
2022-10-16T01:08:59+08:00:     amazon-ebs: +++ seq 0 3
2022-10-16T01:08:59+08:00:     amazon-ebs: ++ for attempt in '`seq 0 $MAX_RETRIES`'
2022-10-16T01:08:59+08:00:     amazon-ebs: ++ rc=0
2022-10-16T01:08:59+08:00:     amazon-ebs: ++ [[ 0 -gt 0 ]]
2022-10-16T01:08:59+08:00:     amazon-ebs: ++ aws ecr get-login-password --region us-west-2
2022-10-16T01:08:59+08:00:     amazon-ebs:
2022-10-16T01:08:59+08:00:     amazon-ebs: Unable to locate credentials. You can configure credentials by running "aws configure".
2022-10-16T01:08:59+08:00:     amazon-ebs: ++ rc=253
2022-10-16T01:08:59+08:00:     amazon-ebs: ++ [[ 253 -eq 0 ]]
2022-10-16T01:08:59+08:00:     amazon-ebs: ++ [[ 0 -eq 3 ]]
2022-10-16T01:08:59+08:00:     amazon-ebs: ++ local jitter=7
2022-10-16T01:08:59+08:00:     amazon-ebs: ++ local sleep_sec=17
2022-10-16T01:08:59+08:00:     amazon-ebs: ++ sleep 17
2022-10-16T01:09:16+08:00:     amazon-ebs: ++ for attempt in '`seq 0 $MAX_RETRIES`'
2022-10-16T01:09:16+08:00:     amazon-ebs: ++ rc=0
2022-10-16T01:09:16+08:00:     amazon-ebs: ++ [[ 1 -gt 0 ]]
2022-10-16T01:09:16+08:00:     amazon-ebs: ++ echo 'Attempt 1 of 3'
2022-10-16T01:09:16+08:00:     amazon-ebs: Attempt 1 of 3
2022-10-16T01:09:16+08:00:     amazon-ebs: ++ aws ecr get-login-password --region us-west-2
2022-10-16T01:09:17+08:00:     amazon-ebs:
2022-10-16T01:09:17+08:00:     amazon-ebs: Unable to locate credentials. You can configure credentials by running "aws configure".
2022-10-16T01:09:17+08:00:     amazon-ebs: ++ rc=253
2022-10-16T01:09:17+08:00:     amazon-ebs: ++ [[ 253 -eq 0 ]]
2022-10-16T01:09:17+08:00:     amazon-ebs: ++ [[ 1 -eq 3 ]]
2022-10-16T01:09:17+08:00:     amazon-ebs: ++ local jitter=2
2022-10-16T01:09:17+08:00:     amazon-ebs: ++ local sleep_sec=22
2022-10-16T01:09:17+08:00:     amazon-ebs: ++ sleep 22
2022-10-16T01:09:39+08:00:     amazon-ebs: ++ for attempt in '`seq 0 $MAX_RETRIES`'
2022-10-16T01:09:39+08:00:     amazon-ebs: ++ rc=0
2022-10-16T01:09:39+08:00:     amazon-ebs: ++ [[ 2 -gt 0 ]]
2022-10-16T01:09:39+08:00:     amazon-ebs: ++ echo 'Attempt 2 of 3'
2022-10-16T01:09:39+08:00:     amazon-ebs: Attempt 2 of 3
2022-10-16T01:09:39+08:00:     amazon-ebs: ++ aws ecr get-login-password --region us-west-2
2022-10-16T01:09:39+08:00:     amazon-ebs:
2022-10-16T01:09:39+08:00:     amazon-ebs: Unable to locate credentials. You can configure credentials by running "aws configure".
2022-10-16T01:09:40+08:00:     amazon-ebs: ++ rc=253
2022-10-16T01:09:40+08:00:     amazon-ebs: ++ [[ 253 -eq 0 ]]
2022-10-16T01:09:40+08:00:     amazon-ebs: ++ [[ 2 -eq 3 ]]
2022-10-16T01:09:40+08:00:     amazon-ebs: ++ local jitter=6
2022-10-16T01:09:40+08:00:     amazon-ebs: ++ local sleep_sec=46
2022-10-16T01:09:40+08:00:     amazon-ebs: ++ sleep 46
2022-10-16T01:10:26+08:00:     amazon-ebs: ++ for attempt in '`seq 0 $MAX_RETRIES`'
2022-10-16T01:10:26+08:00:     amazon-ebs: ++ rc=0
2022-10-16T01:10:26+08:00:     amazon-ebs: ++ [[ 3 -gt 0 ]]
2022-10-16T01:10:26+08:00:     amazon-ebs: ++ echo 'Attempt 3 of 3'
2022-10-16T01:10:26+08:00:     amazon-ebs: Attempt 3 of 3
2022-10-16T01:10:26+08:00:     amazon-ebs: ++ aws ecr get-login-password --region us-west-2
2022-10-16T01:10:26+08:00:     amazon-ebs:
2022-10-16T01:10:26+08:00:     amazon-ebs: Unable to locate credentials. You can configure credentials by running "aws configure".
2022-10-16T01:10:26+08:00:     amazon-ebs: ++ rc=253
2022-10-16T01:10:26+08:00:     amazon-ebs: ++ [[ 253 -eq 0 ]]
2022-10-16T01:10:26+08:00:     amazon-ebs: ++ [[ 3 -eq 3 ]]
2022-10-16T01:10:26+08:00:     amazon-ebs: ++ return exit 253
2022-10-16T01:10:26+08:00:     amazon-ebs: /etc/eks/containerd/pull-image.sh: line 17: return: exit: numeric argument required
2022-10-16T01:10:26+08:00:     amazon-ebs: + ecr_password=
2022-10-16T01:10:26+08:00:     amazon-ebs: + retry sudo ctr --namespace k8s.io image pull 602401143452.dkr.ecr.us-west-2.amazonaws.com/eks/pause:3.5 --user AWS:
2022-10-16T01:10:26+08:00:     amazon-ebs: + local rc=0
2022-10-16T01:10:26+08:00:     amazon-ebs: ++ seq 0 3
2022-10-16T01:10:26+08:00:     amazon-ebs: + for attempt in '`seq 0 $MAX_RETRIES`'
2022-10-16T01:10:26+08:00:     amazon-ebs: + rc=0
2022-10-16T01:10:26+08:00:     amazon-ebs: + [[ 0 -gt 0 ]]
2022-10-16T01:10:26+08:00:     amazon-ebs: + sudo ctr --namespace k8s.io image pull 602401143452.dkr.ecr.us-west-2.amazonaws.com/eks/pause:3.5 --user AWS:
```

Need to add a line for checking `$ecr_password` existence before trying to pull images.

```
[[ -z ${ecr_password} ]] && exit 1
```

Also, need to add `timeout XXX` for `ctr image pull ...` or it will hang here, can do nothing but CTRL-C.

Reply:
In addition to addressing: https://github.com/awslabs/amazon-eks-ami/pull/938/files/9e7c7686efb4cfc242db4da0ce7708ffbcaa9708#r998541356

I think adding the sanity check on `ecr_password` is a good idea. I don't know if adding a timeout on `ctr pull` is necessary, because the "hang" is happening because no password was specified in the `--user` value, and `ctr` will fallback to tty input in this case.

Reply:
agree... it should only happen whenever `$ecr_password` is empty.
... if `$ecr_password` is retrieve from aws cli, do we need to check for other exception? I'm not sure about that.

Reply:
Agreed on the empty ecr_password case. The empty case should be the only thing we need to guard though since any other input would result in a response, either success or failure for some reason (unauthorized, invalid, etc). I think the response itself would provide enough information for the user to fix it if they look in the right place.